src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int lastIndex = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                lastIndex--;
+                lastSegment = segments[lastIndex];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, lastIndex);
 
             // Remove the unnecessary type cast segment.
-            ODataPathSegment nextToLastSegment = segments.Last();
-            while (nextToLastSegment is TypeSegment)
+            lastIndex = segments.Count - 1;

```

Comment:
>segments.Count [](start = 24, length = 14)

do we need to recompute lastIndex? isn't it already segments.Count - 1 (appropriately adjusted for any nav prop/operation segments above)?